scrapy/core/spidermw.py

```diff
@@ -60,11 +60,22 @@ def process_spider_exception(_failure):
             return _failure
 
         def process_spider_output(result):
+            def wrapper(result_iterable):
+                try:
+                    for r in result_iterable:
+                        yield r
+                except Exception as ex:
+                    exception_result = process_spider_exception(Failure(ex))
+                    if exception_result is None or isinstance(exception_result, Failure):
+                        raise
+                    for output in exception_result:
+                        yield output
             for method in self.methods['process_spider_output']:
                 result = method(response=response, result=result, spider=spider)
-                assert _isiterable(result), \
-                    'Middleware %s must returns an iterable object, got %s ' % \
-                    (fname(method), type(result))
+                if _isiterable(result):
```

Comment:
I think it doesn't make sense to catch this error in process_spider_exception - this is a programmer error, not a problem with data. So it'd be nice if `process_spider_exception` still won't be able to catch this error.

Another question: what happens if an error from `process_spider_output` is handled by `process_spider_exception`? For `process_spider_input` or for spider itself it is documented that `process_spider_output` chain is invoked, but for process_spider_exception it doesn't make sense to start process_spider_output from the beginning. 

I think executing `process_spider_output` methods which are not executed yet makes sense in this case, but this should be tested and documented.